docs/book/how-to/handle-data-artifacts/datasets.md

```diff
@@ -0,0 +1,171 @@
+
+## Introduction to Custom Dataset Classes
+
+Custom Dataset classes in ZenML provide a way to encapsulate data loading, processing, and saving logic for different data sources. They're particularly useful when:
+
+1. Working with multiple data sources (e.g., CSV files, databases, cloud storage)
+2. Dealing with complex data structures that require special handling
+3. Implementing custom data processing or transformation logic
+
+## Implementing Dataset Classes for Different Data Sources
+
+Let's create a base Dataset class and implement it for CSV and BigQuery data sources:
+
+```python
+from abc import ABC, abstractmethod
+import pandas as pd
+from google.cloud import bigquery
+from typing import Optional
+
+class Dataset(ABC):
+    @abstractmethod
+    def read_data(self) -> pd.DataFrame:
+        pass
+
+class CSVDataset(Dataset):
+    def __init__(self, data_path: str, df: Optional[pd.DataFrame] = None):
+        self.data_path = data_path
+        self.df = df
+
+    def read_data(self) -> pd.DataFrame:
+        if self.df is None:
+            self.df = pd.read_csv(self.data_path)
+        return self.df
+
+class BigQueryDataset(Dataset):
+    def __init__(
+        self,
+        table_id: str,
+        df: Optional[pd.DataFrame] = None,
+        project: Optional[str] = None,
+    ):
+        self.table_id = table_id
+        self.project = project
+        self.df = df
+        self.client = bigquery.Client(project=self.project)
+
+    def read_data(self) -> pd.DataFrame:
+        query = f"SELECT * FROM `{self.table_id}`"
+        self.df = self.client.query(query).to_dataframe()
+        return self.df
+
+    def write_data(self) -> None:
+        job_config = bigquery.LoadJobConfig(write_disposition="WRITE_TRUNCATE")
+        job = self.client.load_table_from_dataframe(self.df, self.table_id, job_config=job_config)
+        job.result()
+```
+
+## Creating Custom Materializers
+
+[Materializers](./handle-custom-data-types.md) in ZenML handle the serialization and deserialization of artifacts. Custom Materializers are essential for working with custom Dataset classes:
+
+```python
+from typing import Type
+from zenml.materializers import BaseMaterializer
+from zenml.io import fileio
+from zenml.enums import ArtifactType
+import json
+import os
+
+class CSVDatasetMaterializer(BaseMaterializer):
+    ASSOCIATED_TYPES = (CSVDataset,)
+    ASSOCIATED_ARTIFACT_TYPE = ArtifactType.DATA
+
+    def load(self, data_type: Type[CSVDataset]) -> CSVDataset:
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "r") as f:
+            metadata = json.load(f)
+        return CSVDataset(metadata["data_path"])
+
+    def save(self, dataset: CSVDataset) -> None:
+        metadata = {"data_path": dataset.data_path}
+        with fileio.open(os.path.join(self.uri, "metadata.json"), "w") as f:
+            json.dump(metadata, f)
+        if dataset.df is not None:
```

Comment:
Shouldn't this call `dataset.read_data()` which returns a dataframe, and then save that one?

Also, by having this path somehow on the dataset, it sort of goes against ZenML materializers. This csv file is most likely not stored in the artifact store but locally somewhere, which means this materializer is useless in remote scenarios.

Reply:
Instead, this should IMO be like this:
```python
def save(self, dataset: CSVDataset) -> None:
  data_path = os.path.join(self.uri, "data.csv")
  # now write it to a temp location and copy to the artifact store


def load(self, ...):
  # copy somewhere locally
  return CSVDataset(local_temp_path)
```